### PR TITLE
Avoid using socket for hints in generateCPUTopologyHints

### PR DESCRIPTION
**Pull request by AlexeyPerevalov**
Sockets don't affect performance as NUMA node does, since NUMA node has dedicated memory controller, but socket it's physical extension point.
Socket it's only cpu specific thing and it's strange to merge bitmask of
deviceplugin's and cpu manager, when cpu manager takes into account socket.

/kind cleanup
/sig node

### CODE REVIEW

**Comment by AlexeyPerevalov** [comment]
Previous pull-kubernetes-e2e-gce failed due to:
[{PodScheduled False 0001-01-01 00:00:00 +0000 UTC 2020-05-18 15:20:54 +0000 UTC Unschedulable 0/4 nodes are available: 1 node(s) were unschedulable, 3 node(s) didn't match node selector.}]

/test pull-kubernetes-e2e-gce

**Comment by klueska** [comment]
Hey @AlexeyPerevalov thanks for putting this PR together, and sorry for the delay in reviewing it.

When we discussed this change on slack, I didn't realize you planned on removing the entire Socket abstraction from the CPU allocation algoirthm. I think the notion of `Socket` is still important in the memory hierarchy for determining relative cpu allocations, because it dictates whether two CPUs share the same L2 cache or not. The `Socket` abstraction doesn't matter for cross-device alignment, but it does matter for cpu-specific alignment.

I thought you were only planning on making the following changes, which all look good to me:

1) https://github.com/kubernetes/kubernetes/pull/90980/files#diff-06aa1b61a9bc4a50aa8d8b0a69d95bb2 

2) https://github.com/kubernetes/kubernetes/pull/90980/files#diff-67c46e7cbb41ea5e675000a3298e6b32

3) https://github.com/kubernetes/kubernetes/pull/90980/files#diff-f1223d0d0c42e67041357f172a322ab6

4) https://github.com/kubernetes/kubernetes/pull/90980/files#diff-b54ee8195220ea88e59b6eb423e83dd1

5) https://github.com/kubernetes/kubernetes/pull/90980/files#diff-a19f204f56e91b0871491b88c0e0b139

**Comment by AlexeyPerevalov** [comment]
> Hey @AlexeyPerevalov thanks for putting this PR together, and sorry for the delay in reviewing it.
> 
> When we discussed this change on slack, I didn't realize you planned on removing the entire `Socket` abstraction from the CPU allocation algorithm.
> 
> I think the notion of `Socket` is still important in the memory hierarchy for determining relative cpu allocations, because it dictates whether two CPUs share the same L2 cache or not. The `Socket` abstraction doesn't matter for cross-device alignment, but it does matter for cpu-specific alignment.
> 
> I thought you were only planning on making the following changes, which all look good to me:
> 
> 1. pkg/kubelet/cm/container_manager_linux.go https://github.com/kubernetes/kubernetes/pull/90980/files#diff-06aa1b61a9bc4a50aa8d8b0a69d95bb2
> 2. pkg/kubelet/cm/cpumanager/cpu_manager.go https://github.com/kubernetes/kubernetes/pull/90980/files#diff-67c46e7cbb41ea5e675000a3298e6b32
> 3. pkg/kubelet/cm/cpumanager/policy_static.go https://github.com/kubernetes/kubernetes/pull/90980/files#diff-f1223d0d0c42e67041357f172a322ab6
> 4. pkg/kubelet/cm/devicemanager/manager.go https://github.com/kubernetes/kubernetes/pull/90980/files#diff-b54ee8195220ea88e59b6eb423e83dd1
> 5. pkg/kubelet/cm/topologymanager/topology_manager.go https://github.com/kubernetes/kubernetes/pull/90980/files#diff-a19f204f56e91b0871491b88c0e0b139

Thanks for review, I got you point, I'll keep *socket* untouched )

**Comment by AlexeyPerevalov** [comment]
it requires cadvisor v0.37

**Comment by cbf123** [comment]
@klueska You had said earlier "I think the notion of Socket is still important in the memory hierarchy for determining relative cpu allocations, because it dictates whether two CPUs share the same L2 cache or not."

Just to nitpick, but isn't L2 per-core (split between hyperthreads if they're enabled), and it's L3 that gets split with sub-numa clustering?

Longer-term I wonder if rather than explicitly naming a "Socket" entity, it might make more sense to construct a topology model expressing logical "distance" and try to select CPUs that are "near" to each other.

File: pkg/kubelet/cm/cpumanager/topology/topology.go [hunk 1]
```diff
@@ -218,34 +216,29 @@ func (d CPUDetails) CPUsInCores(ids ...int) cpuset.CPUSet {
...
-						if cset.Contains(cpu) {
-							numaNodeID = id
-						}
-					}
+					numaNodeID := node.Id
 					CPUDetails[cpu] = CPUInfo{
 						CoreID:     coreID,
-						SocketID:   socket.Id,
+						SocketID:   core.SocketID,
 						NUMANodeID: numaNodeID,
```
Comment (cbf123) [inline pkg/kubelet/cm/cpumanager/topology/topology.go:236]: nit: could just use node.Id here, no?

Comment (AlexeyPerevalov) [reply pkg/kubelet/cm/cpumanager/topology/topology.go:236]: ok, maybe numaNodeID is unnecessary here.

**Comment by klueska** [comment]
> Just to nitpick, but isn't L2 per-core (split between hyperthreads if they're enabled), and it's L3 that gets split with sub-numa clustering?

It depends on the architecture. What I probably should have said instead of L2 is that we care about the Last-Level-Cache (LLC), which may the the L2, L3, or even sometimes the L4 cache. The point being, that the LLC is a per-socket cache, and an important piece of the memory hierarchy when it comes to deciding which cores are "closest" to one another.

It has nothing to do with sub-numa clustering though. It's quite common to have multiple sockets, all part of a single NUMA domain, which is when this would be most important.

> Longer-term I wonder if rather than explicitly naming a "Socket" entity, it might make more sense to construct a topology model expressing logical "distance" and try to select CPUs that are "near" to each other.

The recommended method (at least on x86) is to use exactly the fields presented in this struct (which includes SocketID) to calculate the CPU topology: https://wiki.osdev.org/Detecting_CPU_Topology_(80x86)

Here is an implementation of the algorithm discussed in that link (which I wrote 5 or so years ago for an open source OS I was working): https://github.com/akaros/akaros/blob/e2e76c34b4eb027eec44bf0ab81523c4359aa6a3/kern/src/corealloc_packed.c.

File: pkg/kubelet/cm/cpumanager/topology/topology.go (hunk 1 above)
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology/topology.go:236]: Agreed, this seems unnecessary

**Review by klueska — APPROVED** [review]
One minor nit, but other than that, this looks great. Thanks for staying on top of things and getting the change to `cadvisor` pushed through!

**Comment by justaugustus** [comment]
@AlexeyPerevalov @kubernetes/sig-node-pr-reviews -- this needs a release note

**Pull request merged**